src/bin/listbox_model.rs

```diff
@@ -0,0 +1,322 @@
+
+#[macro_use]
+extern crate glib;
+extern crate gio;
+extern crate gtk;
+
+extern crate glib_sys as glib_ffi;
+extern crate gobject_sys as gobject_ffi;
+
+extern crate gobject_subclass;
+
+use gio::prelude::*;
+use gtk::prelude::*;
+
+use std::env::args;
+
+use row_data::RowData;
+
+// make moving clones into closures more convenient
+macro_rules! clone {
+    (@param _) => ( _ );
+    (@param $x:ident) => ( $x );
+    ($($n:ident),+ => move || $body:expr) => (
+        {
+            $( let $n = $n.clone(); )+
+            move || $body
+        }
+    );
+    ($($n:ident),+ => move |$($p:tt),+| $body:expr) => (
+        {
+            $( let $n = $n.clone(); )+
+            move |$(clone!(@param $p),)+| $body
+        }
+    );
+}
+
+fn build_ui(application: &gtk::Application) {
+    let window = gtk::ApplicationWindow::new(application);
+
+    window.set_title("ListBox Model Sample");
+    window.set_border_width(10);
+    window.set_position(gtk::WindowPosition::Center);
+    window.set_default_size(320, 480);
+
+    window.connect_delete_event(clone!(window => move |_, _| {
+        window.destroy();
+        Inhibit(false)
+    }));
+
+    let vbox = gtk::Box::new(gtk::Orientation::Vertical, 5);
+
+    let model = gio::ListStore::new(RowData::static_type());
+    let listbox = gtk::ListBox::new();
+    listbox.bind_model(&model, clone!(window => move |item| {
+        let box_ = gtk::ListBoxRow::new();
+        let item = item.clone().downcast::<RowData>().unwrap();
+
+        let hbox = gtk::Box::new(gtk::Orientation::Horizontal, 5);
+
+        let label = gtk::Label::new(None);
+        item.bind_property("name", &label, "label")
+            .flags(glib::BindingFlags::DEFAULT | glib::BindingFlags::SYNC_CREATE)
+            .build();
+        hbox.pack_start(&label, true, true, 0);
+
+        let spin_button = gtk::SpinButton::new_with_range(0.0, 100.0, 1.0);
+        item.bind_property("count", &spin_button, "value")
+            .flags(glib::BindingFlags::DEFAULT | glib::BindingFlags::SYNC_CREATE | glib::BindingFlags::BIDIRECTIONAL)
+            .build();
+        hbox.pack_start(&spin_button, false, false, 0);
+
+        let edit_button = gtk::Button::new_with_label("Edit");
+        edit_button.connect_clicked(clone!(window, item => move |_| {
+            let dialog = gtk::Dialog::new_with_buttons(Some("Edit Item"), Some(&window), gtk::DialogFlags::MODAL,
+                &[("Close", 0)]);
+            dialog.set_default_response(0);
+            dialog.connect_response(|dialog, _| dialog.destroy());
+
+            let content_area = dialog.get_content_area();
+
+            let entry = gtk::Entry::new();
+            item.bind_property("name", &entry, "text")
+                .flags(glib::BindingFlags::DEFAULT | glib::BindingFlags::SYNC_CREATE | glib::BindingFlags::BIDIRECTIONAL)
+                .build();
+            entry.connect_activate(clone!(dialog => move |_| {
+                dialog.response(0);
```

Comment:
@sdroege What does this signal handler do?

Reply:
When you press enter inside the `GtkEntry`, it sends response `0` (see above, "Ok") to the dialog. It is as if you clicked the corresponding button on the dialog

Reply:
Thanks